controllers/oauth-server/index.js

```diff
@@ -0,0 +1,142 @@
+}
+
+function generateIdToken(user) {
+    const idToken = {
+      iss: req.get('host'), // TODO Choose a Partner Domain!!
+      aud: 'https://www.google.com',
+      iat: new Date(),
+      exp: new Date() + ONE_HOUR,
+      sub: user.username,
+      name: user.name
+    }
+
+    return encodeBase64("{\"alg\":\"none\",\"typ\":\"JWT\"}")
+      + "."
+      + encodeBase64(JSON.stringify(authCode))
+      + ".";
+}
+
+router.get('/signin', (request, response) => {
+  const session = request.session;
+  session.signInRequest = {
+        clientId: request.query.client_id,
+        redirectUrl: request.query.redirect_url,
+        state: request.query.state,
+        scope: request.query.scope,
+        responseType: request.query.response_type    
+  };
+
+  response.render('oauth-server/login', {
+    returnUrl: '',
+    readerId: ''
+  });  
+});
+
+router.post('/signin', (request, response, next) => {
+  const signInRequest = request.session.signInRequest;
+  const oAuthClient = oauthClientsModel.findByKey(signInRequest.clientId);
+  if (!oAuthClient) {
+    next('Unknown clientId');
+    return;
+  }
+
+  const authCode = OAuthUtils.generateAuthCode(oAuthClient, user, signinRequest.getScope()); //TODO Add proper User
+  const idToken = OAuthUtils.generateIdToken(user);
+
+  response.redirect("https://oauth-redirect.googleusercontent.com/_/redirect?" +
+          "code=" + authCode +
+          "&id_token=" + idToken +
+          "&state=" + signinRequest.getState());
+});
+
+router.post('/token', (request, response, next) => {
+  const clientId = request.body.client_id;
+  const clientSecret = request.body.client_secret;
+  const oAuthClient = oauthClientsModel.findByKeyAndSecret(clientId, clientSecret);  
+  if (!oAuthClient) {
+    next('Unknown client_id or wrong client_secret');
+  }
+
+  const grantType = request.body.grant_type;
+  switch(grantType) {
+    case "urn:ietf:params:oauth:grant-type:jwt-bearer":
+      doAutoSignInOrUp(oAuthClient, request, response);
+      break;
+    case "authorization_code":
+      doAuthorizationCode(oAuthClient, request, response);
+      break;
+    case "refresh_token":
+      doRefreshToken(oAuthClient, request, response);
+      break;
+    default: 
+      next('Unsupported grant_type: ' + grantType);
+  }  
+});
+
+function doAutoSignInOrUp(oAuthClient, request, response) {
+  const scope = request.getParameter("scope");
+  const assertion = request.getParameter("assertion");
+
+  const googleIdToken = verifier.verify(assertion); // TODO Install Token Verifier library
+  const userEmail = googleIdToken.getPayload().getEmail();
+  const userId = googleIdToken.getPayload().getSubject();
+  // TODO Fetch or Create user from DB
+
+  const authCode = generateAuthCode(oAuthClient, null, scope); // TODO Add Correct User
+  const idToken = generateIdToken(null); //TODO Add Correct User
+
+  const respContent = {
+    authorization_code: authCode,
```

Comment:
I think this should be an access token instead of an auth code. This is consistent with the doc though, so let me check with fargo@.